src/lib/gssapi/mechglue/g_accept_sec_context.c

```diff
@@ -218,25 +220,40 @@ gss_cred_id_t *		d_cred;
 	selected_mech = union_ctx_id->mech_type;
     }
 
+    /* Make a mech selection so we can use it when creating contexts. */
+    mech = gssint_get_mechanism(selected_mech);
+    if (!mech) {
+	status = GSS_S_BAD_MECH;
+	goto error_out;
+    }
+
     /* Now create a new context if we didn't get one. */
     if (*context_handle == GSS_C_NO_CONTEXT) {
-	status = GSS_S_FAILURE;
 	union_ctx_id = (gss_union_ctx_id_t)
-	    malloc(sizeof(gss_union_ctx_id_desc));
+	    calloc(sizeof(gss_union_ctx_id_desc), 1);
 	if (!union_ctx_id)
 	    return (GSS_S_FAILURE);
 
 	union_ctx_id->loopback = union_ctx_id;
 	union_ctx_id->internal_ctx_id = GSS_C_NO_CONTEXT;
+	/* set the new context handle to caller's data */
+	*context_handle = (gss_ctx_id_t)union_ctx_id;
+    }
+
+    if (union_ctx_id->internal_ctx_id == NULL) {
 	status = generic_gss_copy_oid(&temp_minor_status, selected_mech,
-				      &union_ctx_id->mech_type);
+	                              &union_ctx_id->mech_type);
 	if (status != GSS_S_COMPLETE) {
-	    free(union_ctx_id);
-	    return (status);
+	    return status;
 	}
 
-	/* set the new context handle to caller's data */
-	*context_handle = (gss_ctx_id_t)union_ctx_id;
+	if (mech->gss_create_sec_context != NULL) {
+	    status = mech->gss_create_sec_context(
```

Comment:
I think it could even be part of the first commit, unless you find that the code changes are invasive.  It can certainly be part of this PR.